wire-ios-request-strategy/Sources/Message Sending/MessageSenderTests.swift

```diff
@@ -0,0 +1,356 @@
+        )
+
+        let message = GenericMessageEntity(
+            conversation: groupConversation,
+            message: GenericMessage(content: Text(content: "Hello World")),
+            completionHandler: nil)
+
+        let (_, messageSender) = Arrangement(coreDataStack: coreDataStack)
+            .withMessageDependencyResolverCompleting()
+            .withApiVersionResolving(to: .v0)
+            .withSendProteusMessage(returning: .success((messageSendingStatus, response)))
+            .arrange()
+
+        // when
+        let result = await messageSender.sendMessage(message: message)
+
+        // then
+        assertSuccess(result: result)
+    }
+
+    func testThatWhenSendingMessageFailsDueToMissingClients_thenEstablishSessionsAndTryAgain() async throws {
+        // given
+        let response = ZMTransportResponse(payload: nil, httpStatus: 412, transportSessionError: nil, apiVersion: 0)
+        let message = GenericMessageEntity(
+            conversation: groupConversation,
+            message: GenericMessage(content: Text(content: "Hello World")),
+            completionHandler: nil)
+
+        let (arrangement, messageSender) = Arrangement(coreDataStack: coreDataStack)
+            .withMessageDependencyResolverCompleting()
+            .withApiVersionResolving(to: .v0)
+            .withSendProteusMessageFailing(with: NetworkError.missingClients(
+                Arrangement.Scaffolding.messageSendingStatusMissingClients,
+                response)
+            )
+            .withEstablishSessions(returning: .success(Void()))
+            .arrange()
+
+        // when
+        let result = await messageSender.sendMessage(message: message)
+
+        // then
+        assertSuccess(result: result)
+        XCTAssertEqual(Set(arrayLiteral: Arrangement.Scaffolding.clientID), arrangement.sessionEstablisher.establishSessionWithApiVersion_Invocations[0].clients)
+
+    }
+
+    func testThatWhenSendingMessageFailsWithTemporaryError_thenTryAgain() async throws {
+        // given
+        let response = ZMTransportResponse(payload: nil, httpStatus: 408, transportSessionError: nil, apiVersion: 0)
+        let message = GenericMessageEntity(
+            conversation: groupConversation,
+            message: GenericMessage(content: Text(content: "Hello World")),
+            completionHandler: nil)
+
+        let (arrangement, messageSender) = Arrangement(coreDataStack: coreDataStack)
+            .withMessageDependencyResolverCompleting()
+            .withApiVersionResolving(to: .v0)
+            .withSendProteusMessageFailing(with: NetworkError.errorDecodingResponse(response))
+            .withEstablishSessions(returning: .success(Void()))
+            .arrange()
+
+        // when
+        let result = await messageSender.sendMessage(message: message)
+
+        // then
+        assertSuccess(result: result)
+        XCTAssertEqual(2, arrangement.messageApi.sendProteusMessageMessageConversationID_Invocations.count)
```

Comment:
You could actually call the throwing method `.get()`.
Then the `assertSuccess` helper is not needed.

```suggestion
        // when
        try await messageSender.sendMessage(message: message).get()

        // then
        XCTAssertEqual(2, arrangement.messageApi.sendProteusMessageMessageConversationID_Invocations.count)
```

Reply:
True, this all got re-written when I switched to throw based error handling in https://github.com/wireapp/wire-ios/pull/648